### DISCUSSION TITLE
Louvre on a hybrid i915/nouveau system

### DISCUSSION THREAD

**Discussion by yurikhan** [CuarzoSoftware/Louvre#27] [Tested Hardware & Platforms]
I have built Louvre (d69b475) on Ubuntu 24.04 (development). My laptop has hybrid graphics, integrated Intel GPU (i915 driver) and NVIDIA (nouveau). Intel drives the built-in 3840×2160 panel, and NVIDIA drives an external 5120×2880 monitor over (mini) DisplayPort. (Also has an HDMI output, but I wasn’t able to achieve 5K over that, so not using it.) On X11, both monitors give me 60 FPS.

When I start Firefox and navigate to <https://www.testufo.com/>, under louvre-views, it displays an exorbitant FPS value fluctuating between 150 and 512. The animation speed also fluctuates.

The expected behavior is a steady 60 FPS animation on both displays. (Frustratingly, other Wayland compositors, including GNOME/Mutter, Sway/wlroots and Hyprland/wlroots, give me 60 FPS on the built-in panel and about 15–20 FPS on the external monitor.)

Does that look like some unimplemented feature or protocol in Louvre, or am I doing something wrong?

**Comment by ehopperdietzel** [comment]
Interesting, I tried it on my computer a moment ago, and the same thing happens, but only in Firefox (300 fps avg). From Google Chrome, it shows 60 fps. Could you try it on Google Chrome? Or another browser? I will analyze what might be happening.

**Comment by ehopperdietzel** [comment]
Alright, I've identified the issue. The problem stems from both Firefox and Louvre. The issue lies in the fact that Firefox sends multiple [wl_surface::frame](https://wayland.app/protocols/wayland#wl_surface:request:frame) requests in succession and before a wl_surface commit, without waiting for a response from the compositor. As a result, Louvre responds to ALL of these frames after rendering the window and then waits for Firefox to send another one. I've made a modification so that Louvre now responds to only one frame at a time, achieving 60 fps. 

I implemented this partly for security reasons, as if a client generates frames at a rate higher than they are destroyed, it could potentially exhaust the compositor's memory (in an extreme case). Therefore, my solution will be to respond one at a time but impose a limit on how many can live simultaneously. If it exceeds this limit, Louvre will begin to respond/destroy more than one frame at a time.

Thank you for noticing btw. I'll upload the corrected version, most likely tomorrow or the day after, in release 1.2.0-1.

**Reply by yurikhan** [reply]
Nice! Does a bug need to be filed against Firefox? Will you or should I?

**Reply by ehopperdietzel** [reply]
I don't think so. I mean, it's strange what it does, but technically, the protocol description is ambiguous enough to allow this. Also, if it works as expected in Weston or Sway, then the one who needs to change is Louvre.

**Reply by ehopperdietzel** [reply]
By the way, how does your Nvidia card perform with Louvre?

**Reply by yurikhan** [reply]
> By the way, how does your Nvidia card perform with Louvre?

I have not yet collected enough data.

Basically, I’m just toe-dipping Wayland over the New Year holidays to see if it’s in any way feasible. My stable environment is i3/X11/Ubuntu 22.04.

In my workflow, I use the external monitor/keyboard/mouse extensively, with the built-in laptop panel acting as auxiliary (and built-in keyboard/touchpad disabled so that my cats can step or sleep on them without disrupting my work). So migrating to a mainstream compositor that gives me 20 fps on my main output is a non-starter.

If Louvre with the fix for this issue achieves 60 fps, I might try prototyping a tiling compositor (won’t be big and professional like sway!) and possibly wait for xwayland support.

**Reply by ehopperdietzel** [reply]
Alright, well, I don't think I'll be able to release 1.2.0 on time for tomorrow (I am working on fractional scaling). If you want to test it right now, you could try the "fractional" branch. I've fixed the Firefox issue there. Additionally, I plan to add XWayland support for v2.0.0. The API is going to change a bit so in the meantime, I've been adding features to 1.x.x that don't break the API.

**Reply by yurikhan** [reply]
Looks like some unrelated changes got committed in acfb8d7. Depending on how clean you want your history, you might want to revise that.

**Reply by yurikhan** [reply]
On the `fractional` branch, I get 60 FPS as long as the Firefox window is not wider than about 4096 physical pixels wide. After that, it steadily declines to 30 FPS at full width.

If I tile two windows side by side, and one of them shows static content, then the other gets 60 FPS. But if both are updating, 30 FPS again. (Tested with UFO Test on the left and `ls -lR /` on the right; in a real world scenario, left would be the application I’m developing and right would be its logs whooshing by in the terminal.)

That’s on nouveau. With nvidia-535 or nvidia-545, I am unable to start Louvre at all. If the external monitor is connected, it says something about not being able to initialize output DisplayPort-1 and then freezes until killed (via ssh, as local keyboard is unresponsive); if I disconnect the cable before starting Louvre, the `nvidia-drm` kernel module starts spamming my syslog with `[drm:__nv_drm_gem_nvkms_memory_prime_get_sg_table [nvidia_drm]] *ERROR* [nvidia-drm] [GPU ID 0x00000100] Cannot create sg_table for NvKmsKapiMemory 0x00000000844ebc17`, CPU usage at 100% (of one core), again, until louvre-views is killed.

**Reply by ehopperdietzel** [reply]
Okay, thank you! What seems to be happening is that your integrated GPU is likely handling all the rendering, while the Nvidia card is merely displaying the buffer rendered by the integrated GPU. Ideally, each GPU should render its own connectors, but for that to occur, they need to share buffers (application textures) through Direct Memory Access (DMA). For some reason, I've observed that Nouveau drivers often encounter issues when importing buffers, they import them but with an offset, resulting in a glitchy appearance. Due to this, I've currently disabled that feature.

Having said that, I'm considering adding an environment variable to the SRM backend to provide the option of enabling it (maybe it works just fine in your case). Additionally, I'm thinking of incorporating the ability to choose the primary GPU, which is currently selected automatically.

Regarding Nvidia proprietary drivers, I'm unsure about the recent issues causing the mentioned bug. Even my own card isn't working properly; it starts at 60 fps and then gradually drops, with an increase in CPU usage. I haven't changed the driver, so I suspect it might be something related to the Kernel or perhaps some driver configuration.

**Reply by ehopperdietzel** [reply]
I've just added a new environment variable to SRM to override the default allocator device (the GPU that creates the buffers). You can set SRM_ALLOCATOR_DEVICE=/dev/dri/card1, for example, assuming that /dev/dri/card1 represents your Nvidia GPU, allowing it to render to its own connector.

I tested it on a notebook with 2 GPUs (Intel + Nvidia), and surprisingly, it works! The Nvidia GPU creates the buffers, and the Intel GPU imports them through DMA, enabling both GPUs to render their own content independently.

If you'd like to try it, you can find it in the "experimental-env" branch at https://github.com/CuarzoSoftware/SRM/tree/experimental-env

**Reply by yurikhan** [reply]
Hm, no dice. `SRM_ALLOCATOR_DEVICE=/dev/dri/card1 louvre_views` seems to successfully initialize both displays, but all they display is a white topbar (no logo, no text, no clock) and a black background.

I also found the `SRM_DEBUG` variable and collected logs in both the medium-good and bad case.

In the medium-good case, the interesting lines are:

```
SRM error: srmBufferGetTextureID failed. Target device (/dev/dri/card1) has not the PRIME import cap.                                                 
SRM error: srmBufferGetTextureID error. Failed to create EGL image.                                                                                   
SRM warning: Failed o get texture ID from buffer 0 on device /dev/dri/card1 connector 103 (DUMB MODE).                                                
```

I take it that means it fails to share buffers the efficient way, and resorts to copying?

In the bad case, this message is repeated many times:

```
SRM warning: [/dev/dri/card1] Directly mapping buffer DMA fd failed. Trying gbm_bo_map.
```

[louvre-bad.log](https://github.com/CuarzoSoftware/Louvre/files/13842453/louvre-bad.log)
[louvre-good.log](https://github.com/CuarzoSoftware/Louvre/files/13842454/louvre-good.log)

**Reply by ehopperdietzel** [reply]
Oh :( it appears that buffer allocation is failing in your case. Are you using nouveau or nvidia-drm? Additionally, could you please provide the log from running drm_info or srm-display-info? Thank you!

**Reply by yurikhan** [reply]
Nouveau. I described what happens with nvidia-drm [a few comments back](https://github.com/CuarzoSoftware/Louvre/discussions/27#discussioncomment-8011395).

[drm_info.log](https://github.com/CuarzoSoftware/Louvre/files/13843902/drm_info.log)
[srm-display-info.log](https://github.com/CuarzoSoftware/Louvre/files/13843905/srm-display-info.log)

**Reply by yurikhan** [reply]
Something bothers me here. srm-display-info says `/dev/dri/card1` lacks `primeImport` and `primeExport` capabilities, but those are supposed to be [initialized](https://github.com/CuarzoSoftware/SRM/blob/44b7dc9c4b6f9cb4aac8e35971985d39400c8ac9/src/lib/private/SRMDevicePrivate.c#L562) from the result of `drmGetCap(…, DRM_CAP_PRIME, …)` and that call is supposed to always return `DRM_PRIME_CAP_IMPORT | DRM_PRIME_CAP_EXPORT` in Linux 6.6.0+ and `drm_info` in fact says it does.

Aha, you mentioned earlier you had [disabled PRIME on Nouveau](https://github.com/CuarzoSoftware/SRM/blob/44b7dc9c4b6f9cb4aac8e35971985d39400c8ac9/src/lib/private/SRMCorePrivate.c#L220-L224). I should probably try re-enabling that for my testing.

**Reply by yurikhan** [reply]
Ha! Did that. Now the FPS on the full-width Firefox displaying UFO Test is fluctuating from about 33 to about 50.

Also now I understand what you meant by “Nouveau drivers often encounter issues when importing buffers, they import them but with an offset, resulting in a glitchy appearance”.

![IMG_2024-01-05-23-30-00-887](https://github.com/CuarzoSoftware/Louvre/assets/238691/8ba95911-0df2-4ccf-9bb0-959f5481331c)

I recognize this kind of glitch. It can happen when the writer of an image aligns each row of pixels to memory addresses that are a multiple of a certain number of bytes, leaving unused padding pixels or bytes between the end of row Y and start of row Y+1, but the reader interprets the image as contiguous rows of pixels.

I also observe that as I resize the window the slope of the glitch changes. I think if I get the window width to be a multiple of maybe 4 pixels then it displays straight. That is exactly what would happen if an RGB24-formatted image aligned each row to a multiple of 4 bytes using padding pixels.

The technical term for the memory address difference from the start of row Y to start of row Y+1 is called *stride*. Are you sure you’re correctly handling and/or forwarding strides everywhere?

**Reply by ehopperdietzel** [reply]
Great haha! The same happens to me when Nouveau imports DMA buffers. Even if the PRIME cap flags indicate a value of 3, SRM conducts a brief test to determine if it can import/export buffers with GBM/EGL. This test sometimes fails, even when it claims to support it (I believe Nouveau has an option to enable/disable GBM/EGL support during compilation). Consequently, those values may get overridden. Specifically for Nouveau, I had disabled that feature due to this bug, but using the SRM_ALLOCATOR_DEVICE env should no longer disable it by default.

Now, regarding Firefox, I believe it uses the [linux-dma-buf](https://wayland.app/protocols/linux-dmabuf-unstable-v1) Wayland protocol, so it defines the DMA buffer parameters like width, height, stride, offset, etc. Louvre then passes these parameters to SRM using srmBufferCreateFromDMA(), and internally, this function uses [this EGL extension](https://registry.khronos.org/EGL/extensions/EXT/EGL_EXT_image_dma_buf_import.txt) for converting it to an OpenGL texture.

The thing is that if the same DMA parameters are passed to the i915 and Nouveau drivers, and the former successfully imports them while the latter fails (displaying that glitch), then the issue seems to lie with Nouveau, unless it requires something else, such as an additional offset or similar which I don't know.

**Reply by ehopperdietzel** [reply]
If you would like to try hacking SRM to see if you can import them correctly in Nouveau, the functions where this is handled are `srmBufferCreateFromDMA()` and `srmBufferGetTextureID()`. Essentially, `srmBufferCreateFromDMA()` only defines the parameters of the buffer. Then, when `srmBufferGetTextureID()` is called, depending on the GPU, it creates the specific OpenGL texture for that GPU or simply returns it if it has already been created before. So that is where the DMA importing takes place.

https://github.com/CuarzoSoftware/SRM/blob/main/src/lib/SRMBuffer.c

**Reply by yurikhan** [reply]
I added debug logging in `srmBufferCreateFromDMA`, `srmBufferCreateFromCPU`, and `srmBufferGetTextureID`, just to see what real world data looks like. Some observations:

* All buffers I have seen had a single plane.
* Buffers created from CPU have rows aligned to 256 bytes.
* Buffers created from DMA have rows aligned to 4 bytes (one physical pixel).
* If I map the data into memory and dump the pixel data to files before the call to `eglCreateImage`, the data appears to correspond to the declared stride.
* When I resize a window by dragging its edges with my mouse, the smallest increment I can achieve is 2 physical pixels. I suspect resizing happens in 1-unit increments in surface coordinates, which are then scaled.
* On the external screen, a window is displayed correctly if its buffer width is a multiple of 8 physical pixels (and its stride a multiple of 32 bytes). Otherwise, it is distorted at angles of 2:1, 4:1 or 8:1.
* I found [a 6-year-old Mesa-dev thread](https://www.mail-archive.com/mesa-dev@lists.freedesktop.org/msg191822.html) discussing a similar issue.
* I also tried hacking the Louvre code to restrict resizing windows to 4-unit multiples. Unfortunately, that part of the code works in surface units and window geometry, and the full buffer size comes out a bit unpredictable. For some windows, it gets me correct rendering; for others, consistent distorsion. I’m not currently sure how to fudge window sizes in the compositor so that texture strides come out as multiples of 32 bytes (considering fractional scaling), and if popups will pose an additional problem.

Where do DMA buffers come from? Does the client allocate them using some well-known API? Maybe that API can be persuaded to overalign?

**Reply by ehopperdietzel** [reply]
Wow, you went above and beyond with your research. If what you're saying is accurate, then it sounds like bad news. As you mentioned, the only hope would be to compel the client to always use a stride with that multiple. I'm not exactly sure how Firefox creates those buffers, but it's likely using the wayland-egl platform, which I think uses the wl_drm protocol to authenticate an fd and obtain permissions to allocate buffers (I assume with that fd wayland-egl makes the same calls to the driver as GBM does). Then, if the linux-dma-buff protocol is also present, it uses it to share the fd and buffer parameters with the compositor.

If you look at the API for clients, you can see that it has no control over the stride, so it must be calculated internally by the implementation:

[https://gitlab.freedesktop.org/wayland/wayland/-/blob/main/egl/wayland-egl-core.h](https://gitlab.freedesktop.org/wayland/wayland/-/blob/main/egl/wayland-egl-core.h)

Now, this is good because there could be some environment variable or configuration that allows forcing a specific stride. But if you look at the implementation, it must depend on the specific driver used:

[https://gitlab.freedesktop.org/wayland/wayland/-/tree/main/egl](https://gitlab.freedesktop.org/wayland/wayland/-/tree/main/egl)

In my case, the buffers are displayed correctly, probably because they are created by the Nouveau driver and then imported by i915, which apparently has no issues. However, in your case, it seems that Nouveau does not support creating DMA buffers, or at least the GBM implementation (you mentioned everything turns black when Nouveau is the allocator device), but maybe it can be enabled if you compile it manually.

**Reply by yurikhan** [reply]
I’m getting the impression that clients just call GBM directly, over one of the devices in `/dev/dri`. Firefox [has an environment variable](https://searchfox.org/mozilla-central/source/widget/gtk/DMABufLibWrapper.cpp#185) to override which one it uses, which I have not yet tested.

I’ve looked into Mesa’s libgbm and it looks like it lets the user to specify a backend library via the `GBM_BACKEND` environment variable. It might be possible to implement one that wraps the Mesa implementation.

Although it’s not just Firefox that has a problem. Nautilus (the GNOME file manager) also gets distorted on the external screen, and I’m pretty sure it uses stock GTK+. So many dependencies to chase :sweat_smile:

**Comment by yurikhan** [comment]
> In the bad case [`SRM_ALLOCATOR_DEVICE=/dev/dri/card1`], this message is repeated many times:
> 
> ```
> SRM warning: [/dev/dri/card1] Directly mapping buffer DMA fd failed. Trying gbm_bo_map.
> ```

This message is written from two places in SRMBuffer.c, one in `CreateFromCPU`, the other in `CreateFromGBM`. With my additional logging, I notice that the one that is firing is in `CreateFromCPU`.

Also, when a syscall or a libc function fails, they usually set `errno`. I logged that and it turns out that the two `mmap` calls fail with `EACCES` (Permission denied). Does that tell you anything?

**Reply by ehopperdietzel** [reply]
Hmm strange, can you try jumping directly to `fail:` [here](https://github.com/CuarzoSoftware/SRM/blob/main/src/lib/private/SRMBufferPrivate.c) ? (in `srmBufferGetDMAFDFromBO()`)

**Reply by ehopperdietzel** [reply]
And another idea, can you try opening Firefox or weston-simple-egl from the compositor with Nouveau as the allocator? Maybe they can successfully create DMA buffers, and what is failing is SRM when creating buffers from main memory.

**Reply by yurikhan** [reply]
So…

By “jumping directly to `fail:`”, do you mean skipping the PRIME export path, straight to `gbm_bo_get_fd`? I don’t think it should make any difference as control reaches there anyway with the current `mmap` failures.

With Nouveau as the allocator, I can open the terminal and from there Nautilus, but I cannot see any of their outputs. I only know they start from the log.

I extracted the logic of creating buffers (via `gbm_bo_create_with_modifiers`), exporting them (via ioctl `DRM_IOCTL_PRIME_HANDLE_TO_FD`) and mapping the resulting fd via `mmap`. As a standalone program, that works (as in, all calls succeed up to and including `mmap`), with any of `/dev/dri/card0`, `/dev/dri/card1`, `/dev/dri/renderD128`, or `/dev/dri/renderD129` as the gbm device. Logging all the calls’ arguments, I cannot see a difference, however, in Louvre/SRM the `mmap` fails with `EACCES`. I will continue trying to get a minimum failing example.

(I also note in `srmBufferGetDMAFDFromBO` you try to enable the `O_RDWR` access mode on the fd returned from PRIME export. However, both the Linux manual and the POSIX spec for `fcntl` tell me `F_SETFL` will not change access mode bits. Does this part ever have any effect?)

**Reply by ehopperdietzel** [reply]
Are you opening devices with `open()` and running the tests with elevated permissions? Louvre by default relies on libseat for opening DRM devices, so maybe this is causing issues. You can turn off libseat with **LOUVRE_ENABLE_LIBSEAT**=0 (though TTY switching won't work) or try the `srm-all-connectors` example, which doesn't use libseat. As for the `fcntl` call, for some reason it is required to make buffers mappable on a small ARM device I have with a Mali GPU using the Lima driver.

**Reply by yurikhan** [reply]
I was opening `/dev/dri/card#` with `open()`, but I was running as my normal user. Now I added libseat (copying from its example) and the test program still works.

(Now if I run the test program from within an existing X11 session, it gets into a weird state: the laptop panel switches to text mode, Ctrl+Alt+F[1-7] switches it between ttys, and keyboard input goes to *both* the active tty and the X11 session still displaying on the external monitor. I understand this should not be expected to “work”, I was curious to see how the test program fails in this case, but it just happily wreaks my X.)

For the record, my test system boots to gdm (GNOME display manager) on tty1. I then switch to tty2 with Ctrl+Alt+F2, log in as my normal user, start `tmux`, under that `mc`, and from there either the test program or Louvre.

My next step will be to compare how Louvre uses libseat against the example code.

**Reply by ehopperdietzel** [reply]
So definitely the issue must be with SRM, can you tell me specifically which distribution you are using? If I try the same one, maybe I can replicate the error here and solve it.

**Reply by ehopperdietzel** [reply]
It's also weird that elevated privileges aren't required to write to those devices. Typically, only the root user or members of the video group can do that. Could you share the output you get by running `ls -l /dev/dri`?

**Reply by yurikhan** [reply]
Sorry for long response times, I’m back to work and finding fewer opportunities to tinker.

My devices in /dev/dri:

```
total 0
drwxr-xr-x   3 root root        140 Jan 21 20:30 .
drwxr-xr-x  22 root root       5580 Jan 21 20:30 ..
drwxr-xr-x   2 root root        120 Jan 21 20:30 by-path
crw-rw----+  1 root video  226,   0 Jan 21 20:30 card0
crw-rw----+  1 root video  226,   1 Jan 21 20:30 card1
crw-rw----+  1 root render 226, 128 Jan 21 20:30 renderD128
crw-rw----+  1 root render 226, 129 Jan 21 20:30 renderD129
```

and I don’t think I’m in the `video` or `render` groups. Despite that, directly calling `open(2)` on these devices succeeds :man_shrugging:

I’m on Ubuntu 24.04.

**Reply by ehopperdietzel** [reply]
No need to worry, perhaps that '+' at the end is granting permissions via ACL. Anyway, when I find the time, I'll test your distribution to see if I can resolve it, and I'll provide you with an update.

**Reply by ehopperdietzel** [reply]
Hi, I still haven't tried to solve the issue, but I just remembered that you can get considerable higher FPS by enabling triple buffering, you can do that by setting **SRM_RENDER_MODE_ITSELF_FB_COUNT**=3 (for displays connected to your Intel card) and **SRM_RENDER_MODE_DUMB_FB_COUNT**=3 for nvidia (since I remember it was using dumb buffers in your case), just use the main SRM branch.
